wgsl/index.bs

```diff
@@ -6411,7 +6426,137 @@ Issue: (dneto) Default rounding mode is an implementation choice.  Is that what
+the Vulkan Memory Model.
+A [SHORTNAME], a [=write access=] is equivalent to a memory write operation in
+the Vulkan Memory Model.
+
+A [=read access=] occurs when an invocation executes one of the following:
+* An evaluation of the [=Load Rule=]
+* Any [[#texture-builtin-functions|texture builtin function]] except:
+    * [[#texturedimensions|textureDimensions]]
+    * [[#texturestore|textureStore]]
+    * [[#texturenumlayers|textureNumLayers]]
+    * [[#texturenumlevels|textureNumLevels]]
+    * [[#texturenumsamples|textureNumSamples]]
+* Any atomic built-in function except [[#atomic-store|atomicStore]]
+
+A [=write access=] occurs when an invocation executes one of the following:
+* An [=assignment statement=]
+* A [[#texturestore|textureStore]] built-in function
+* Any atomic built-in function except [[#atomic-load|atomicLoad]]
+    * [[#atomic-rmw|atomicCompareExchangeWeak]] only performs a write if the
+        second element of the returned result is 1
+
+Read and write accesses do not occur under any other circumstances.
+Read and write accesses are collectively known as [memory
+operations](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-memory-operation)
+in the Vulkan Memory Model.
+
+A memory operation accesses exactly the set of [=memory location|locations=]
+associated with the particular [=memory view=] used in the operation.  For
+example, a memory read that accesses a [=u32=] from a struct containing
+multiple members, only reads the memory locations associated with that u32
+member.
+
+<div class='example memory locations accessed' heading="Accessing memory locations">
+  <xmp>
+    [[block]] struct S {
+      a : f32;
+      b : u32;
+      c : f32;
+    };
+
+    [[group(0), binding(0)]]
+    var<storage> v : S;
+
+    fn foo() {
+      let x = v.b; // Does not access memory locations for v.a or v.c.
+    }
+  </xmp>
+</div>
+
+## Memory Model Reference ## {#memory-model-reference}
+
+Each module-scope variable in [SHORTNAME] forms a unique [memory model
+reference](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-references)
+for the lifetime of a given entry point.
+Each function-scope variable in [SHORTNAME] forms a unique [memory model
+reference](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-references)
+for the lifetime of the variable.
+
+## Scoped Operations ## {#scoped-operations}
+
+[[#atomic-builtin-functions|Atomic built-in functions]] map to [atomic
+operations](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-atomic-operation)
+whose memory
+[scope](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-model-scope)
+is:
+* `Workgroup` if the atomic pointer is in the [=storage classes/workgroup=]
+    storage class
+* `Device`/`QueueFamily` if the atomic pointer is in the [=storage
```

Comment:
Can you explain when it should be `Device` and when it should be `QueueFamily`? These two scopes don't seem to be one and the same from my reading of the Vulkan memory model.

Reply:
Prior to the Vulkan memory model (e.g. Vulkan 1.0 with no extensions), there was no concept of queue family, but device actually meant queue family. So it is queue family, but if you are generating SPIR-V without `OpMemoryModel Logical Vulkan` then you must generate it as `Device`. If you generate SPIR-V with `OpMemoryModel Logical Vulkan` then you would use `QueueFamily`. 

Maybe it would be clearer if I just called it `QueueFamily` and added a note about generating earlier versions of SPIR-V.